lib/linkerd2-trace-context/src/layer.rs

```diff
@@ -0,0 +1,178 @@
+}
+
+// === impl Layer ===
+
+impl<M, S> tower::layer::Layer<M> for Layer<S>
+where
+    S: Clone,
+{
+    type Service = Stack<M, S>;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        Stack {
+            inner,
+            sink: self.sink.clone(),
+        }
+    }
+}
+
+// === impl Stack ===
+
+impl<T, M, S> tower::Service<T> for Stack<M, S>
+where
+    M: tower::Service<T>,
+    S: Clone,
+{
+    type Response = Service<M::Response, S>;
+    type Error = M::Error;
+    type Future = MakeFuture<M::Future, S>;
+
+    fn poll_ready(&mut self) -> Poll<(), M::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let inner = self.inner.call(target);
+
+        MakeFuture {
+            inner,
+            sink: Some(self.sink.clone()),
+        }
+    }
+}
+
+// === impl MakeFuture ===
+
+impl<F: Future, S> Future for MakeFuture<F, S> {
+    type Item = Service<F::Item, S>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = try_ready!(self.inner.poll());
+        let sink = self.sink.take().expect("poll called after ready");
+        Ok(Async::Ready(Service { inner, sink }))
+    }
+}
+
+// === impl Service ===
+
+impl<Svc, B, S> tower::Service<http::Request<B>> for Service<Svc, S>
+where
+    Svc: tower::Service<http::Request<B>>,
+    S: SpanSink + Clone,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = Either<Svc::Future, SpanFuture<Svc::Future, S>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Svc::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut request: http::Request<B>) -> Self::Future {
+        let trace_context = propagation::unpack_trace_context(&request);
+        let mut span = None;
+
+        if let Some(context) = trace_context {
+            trace!(message = "got trace context", ?context);
+            let span_id = propagation::increment_span_id(&mut request, &context);
+            // If we plan to sample this span, we need to record span metadata
+            // from the request before dispatching it to inner.
+            if context.is_sampled() {
+                trace!(message = "span will be sampled", ?span_id);
+                let path = request
+                    .uri()
+                    .path_and_query()
+                    .map(|pq| pq.as_str().to_owned());
+                span = Some(Span {
+                    trace_id: context.trace_id,
+                    span_id,
+                    parent_id: context.parent_id,
+                    span_name: path.unwrap_or_default(),
+                    start: SystemTime::now(),
+                    // End time will be updated when the span completes.
+                    end: SystemTime::now(),
+                });
+            }
+        }
+
+        let f = self.inner.call(request);
+
+        if let Some(span) = span {
+            let span_fut = SpanFuture {
```

Comment:
Again, I'd be reallly inclined to just return a SpanFuture here (always) and make the sink optional, and then handle the optionality in the SpanFuture impl (rather than the either).

But this is just cosmetic

Reply:
My guess is that always returning the same type, rather than using `Either`, may also reduce the impact on build times a bit (since `Either` results in a much longer type), but it's probably a drop in the bucket compared to other things that are making our builds long...

Reply:
This is much more clear to me, personally.  But I can change it if you want.

Reply:
here's my rationalization: as a user of this service/layer, i don't benefit at all from knowing about the eitherness of the response future -- it's all implementation detail. and it's not actually even two different potential futures! there's one (inner) future, and this layer decorates responses with its own logic (optionally doing some things as the future is processed). finally, its idiomatic (at least in our libs and tower) to have a `ResponseFuture` type for each layer (at least when it wraps its inner response future).